File: gamba/wbot.go
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"time"

	"github.com/go-resty/resty/v2"
	"github.com/shopspring/decimal"
)

// Define a struct for the response structure
type Candle struct {
	Timestamp    string          `json:"ts"`
	Open         decimal.Decimal `json:"open"`
	High         decimal.Decimal `json:"high"`
	Low          decimal.Decimal `json:"low"`
	Close        decimal.Decimal `json:"close"`
	Volume       decimal.Decimal `json:"vol"`
	Currency     string          `json:"currency"`
	TimestampEnd string          `json:"timestamp_end"`
	MarketType   string          `json:"market_type"`
}

// Fetch Candles from OKX API
func fetchCandles(symbol string, interval string, limit int) ([]Candle, error) {
	client := resty.New()
	url := "https://www.okx.com/api/v5/market/candles"

	params := map[string]string{
		"instId": symbol,
		"bar":    interval,
		"limit":  fmt.Sprintf("%d", limit),
	}

	resp, err := client.R().SetQueryParams(params).Get(url)
	if err != nil {
		return nil, fmt.Errorf("error fetching candles: %w", err)
	}

	// Log the raw response to debug
	fmt.Println("Raw response:", string(resp.Body()))

	var data struct {
		Data []Candle `json:"data"`
	}

	err = json.Unmarshal(resp.Body(), &data)
	if err != nil {
		return nil, fmt.Errorf("error parsing JSON: %w", err)
	}

	if len(data.Data) == 0 {
		return nil, fmt.Errorf("no candle data retrieved")
	}

	fmt.Println("Fetched candles:", len(data.Data))
	return data.Data, nil
}


// Calculate Moving Averages
func calculateMovingAverages(candles []Candle, maPeriod int) ([]Candle, error) {
	for i := maPeriod - 1; i < len(candles); i++ {
		sum := decimal.NewFromInt(0)
		for j := i - maPeriod + 1; j <= i; j++ {
			sum = sum.Add(candles[j].Close)
		}
		candles[i].Open = sum.Div(decimal.NewFromInt(int64(maPeriod)))
	}
	return candles, nil
}

// Calculate Support and Resistance
func calculateSupportResistance(candles []Candle, window int) ([]Candle, error) {
	for i := window - 1; i < len(candles); i++ {
		minLow := candles[i].Low
		maxHigh := candles[i].High
		for j := i - window + 1; j <= i; j++ {
			if candles[j].Low.Cmp(minLow) < 0 {
				minLow = candles[j].Low
			}
			if candles[j].High.Cmp(maxHigh) > 0 {
				maxHigh = candles[j].High
			}
		}
		// Save the support and resistance
		candles[i].Open = minLow
		candles[i].High = maxHigh
	}
	return candles, nil
}

// Determine Slope of the 125-period MA
func calculateMASlope(candles []Candle, maPeriod int) ([]Candle, error) {
	for i := maPeriod; i < len(candles); i++ {
		// Calculate the slope as the difference between current MA and MA 125 periods ago
		slope := candles[i].Open.Sub(candles[i-maPeriod].Open)
		// Store the slope for later use
		candles[i].Open = slope
	}
	return candles, nil
}

// Calculate Buy/Sell Signals
func calculateCombinedStrategy(candles []Candle, maPeriod int, srWindow int) ([]Candle, error) {
	// Calculate the MA, Support, Resistance, and Slope
	candles, err := calculateMovingAverages(candles, maPeriod)
	if err != nil {
		return nil, err
	}
	candles, err = calculateSupportResistance(candles, srWindow)
	if err != nil {
		return nil, err
	}
	candles, err = calculateMASlope(candles, maPeriod)
	if err != nil {
		return nil, err
	}

	// Generate signals
	for i := 0; i < len(candles); i++ {
		if candles[i].Close.LessThanOrEqual(candles[i].Open) && candles[i].Open.GreaterThan(decimal.NewFromInt(0)) {
			// Buy Signal
			fmt.Println("Buy signal detected")
		} else if candles[i].Close.GreaterThanOrEqual(candles[i].Open) && candles[i].Open.LessThan(decimal.NewFromInt(0)) {
			// Sell Signal
			fmt.Println("Sell signal detected")
		}
	}

	return candles, nil
}

// Place an Order (mocking the place order function)
func placeOrder(symbol string, side string, amount float64) {
	// Mock order placement
	fmt.Printf("Placing %s order: %f %s\n", side, amount, symbol)
}

// Set Leverage (mocking the leverage set function)
func setLeverage(symbol string, leverage int) {
	// Mock leverage setting
	fmt.Printf("Leverage set to %d for %s\n", leverage, symbol)
}

// Trading Bot Logic
func tradingBot(symbols []string, interval string, maPeriod int, srWindow int, tradeAmount float64, leverage int) {
	for _, symbol := range symbols {
		setLeverage(symbol, leverage)
	}

	for {
		for _, symbol := range symbols {
			candles, err := fetchCandles(symbol, interval, 100)
			if err != nil || len(candles) < int(math.Max(float64(maPeriod), float64(srWindow))) {
				fmt.Printf("Insufficient data for %s\n", symbol)
				continue
			}

			candles, err = calculateCombinedStrategy(candles, maPeriod, srWindow)
			if err != nil {
				fmt.Printf("Error calculating strategy for %s: %s\n", symbol, err)
				continue
			}

			latestSignal := candles[len(candles)-1].Open
			if latestSignal.Cmp(decimal.NewFromInt(1)) == 0 {
				// Buy Signal
				placeOrder(symbol, "buy", tradeAmount)
			} else if latestSignal.Cmp(decimal.NewFromInt(-1)) == 0 {
				// Sell Signal
				placeOrder(symbol, "sell", tradeAmount)
			}
		}

		// Sleep for 10 seconds before the next loop
		time.Sleep(10 * time.Second)
	}
}

func main() {
	// List of symbols to trade
	symbols := []string{"SOL-USDT-SWAP", "BTC-USDT-SWAP", "ETH-USDT-SWAP", "DOGE-USDT-SWAP", "ACE-USDT-SWAP"}

	// Run the trading bot
	tradingBot(symbols, "5m",90, 20, 0.5, 15)
}

File: gamba/wbot.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.math.BigDecimal;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.*;
import java.util.concurrent.*;

public class wbot {
    private static final String API_KEY = "4c1029f9-6317-4c87-903f-3923f56de4d1";
    private static final String API_SECRET = "B565C60ED402A2EA0DD41167F60E8CD6";
    private static final String PASS_PHRASE = "Step463656!";
    private static final String API_URL = "https://www.okx.com/api/v5/market/candles";
    private static final HttpClient CLIENT = HttpClient.newHttpClient();
    private static final ObjectMapper objectMapper = new ObjectMapper(); // Jackson ObjectMapper

    public static void main(String[] args) {
        List<String> symbols = Arrays.asList("SOL-USDT-SWAP", "BTC-USDT-SWAP", "ETH-USDT-SWAP", "DOGE-USDT-SWAP", "ACE-USDT-SWAP");
        int maPeriod = 90;
        int srWindow = 20;
        double tradeAmount = 0.5;
        int leverage = 15;

        // Use a thread pool to handle multiple symbols concurrently
        ExecutorService executor = Executors.newFixedThreadPool(symbols.size());

        for (String symbol : symbols) {
            executor.submit(() -> runBot(symbol, maPeriod, srWindow, tradeAmount, leverage));
        }
    }

    private static void runBot(String symbol, int maPeriod, int srWindow, double tradeAmount, int leverage) {
        while (true) {
            try {
                // Fetch candles for the symbol
                List<Candle> candles = fetchCandles(symbol, "5m", 300);
                if (candles.size() < Math.max(maPeriod, srWindow)) {
                    System.out.println("Insufficient data for " + symbol + ", waiting...");
                    Thread.sleep(10000);  // Sleep for 10 seconds before retry
                    continue;
                }

                // Calculate strategy indicators
                List<Candle> updatedCandles = calculateStrategy(candles, maPeriod, srWindow);

                // Get the most recent candle and its corresponding signal
                Candle latestCandle = updatedCandles.get(updatedCandles.size() - 1);
                int signal = latestCandle.signal;

                // Print results for debugging
                System.out.println("Timestamp: " + latestCandle.timestamp);
                System.out.println("Latest signal for " + symbol + ": " + signal);
                System.out.println("Support: " + latestCandle.support + ", Resistance: " + latestCandle.resistance);
                System.out.println("MA Slope: " + latestCandle.sma125Slope);

                // Place an order based on the signal
                if (signal == 1) {
                    placeOrder(symbol, "buy", tradeAmount, leverage);
                } else if (signal == -1) {
                    placeOrder(symbol, "sell", tradeAmount, leverage);
                }

                Thread.sleep(10000);  // Sleep for 10 seconds before fetching the next candle
            } catch (Exception e) {
                System.err.println("Error in trading loop for " + symbol + ": " + e.getMessage());
            }
        }
    }

    private static List<Candle> fetchCandles(String symbol, String interval, int limit) throws Exception {
        String url = API_URL + "?instId=" + symbol + "&bar=" + interval + "&limit=" + limit;
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + API_KEY)
                .build();

        HttpResponse<String> response = CLIENT.send(request, HttpResponse.BodyHandlers.ofString());
        String responseBody = response.body();

        // Parse the response body (using Jackson)
        return parseCandles(responseBody);
    }

    private static List<Candle> parseCandles(String responseBody) {
        List<Candle> candles = new ArrayList<>();
        try {
            // Use Jackson to parse the JSON response
            JsonNode jsonResponse = objectMapper.readTree(responseBody);
            if (jsonResponse.has("data")) {
                JsonNode dataArray = jsonResponse.get("data");

                for (JsonNode candleData : dataArray) {
                    Candle candle = new Candle();
                    candle.timestamp = candleData.get(0).asText();
                    candle.open = new BigDecimal(candleData.get(1).asText());
                    candle.high = new BigDecimal(candleData.get(2).asText());
                    candle.low = new BigDecimal(candleData.get(3).asText());
                    candle.close = new BigDecimal(candleData.get(4).asText());
                    candle.volume = new BigDecimal(candleData.get(5).asText());

                    candles.add(candle);
                }
            }
        } catch (Exception e) {
            System.err.println("Error parsing candle data: " + e.getMessage());
        }
        return candles;
    }

    private static List<Candle> calculateStrategy(List<Candle> candles, int maPeriod, int srWindow) {
        // Calculate moving averages, support, resistance, and signal based on MA slope
        for (int i = maPeriod; i < candles.size(); i++) {
            Candle candle = candles.get(i);
            candle.sma125 = calculateSMA(candles, i, maPeriod);
            candle.sma125Slope = candle.sma125.subtract(candles.get(i - 1).sma125);
            candle.support = calculateSupport(candles, i, srWindow);
            candle.resistance = calculateResistance(candles, i, srWindow);

            // Set signal based on conditions
            if (candle.close.compareTo(candle.support) <= 0 && candle.sma125Slope.compareTo(BigDecimal.ZERO) > 0) {
                candle.signal = 1;  // Buy signal
            } else if (candle.close.compareTo(candle.resistance) >= 0 && candle.sma125Slope.compareTo(BigDecimal.ZERO) < 0) {
                candle.signal = -1;  // Sell signal
            } else {
                candle.signal = 0;  // No signal
            }
        }
        return candles;
    }

    private static BigDecimal calculateSMA(List<Candle> candles, int index, int period) {
        BigDecimal sum = BigDecimal.ZERO;
        for (int i = index - period + 1; i <= index; i++) {
            sum = sum.add(candles.get(i).close);
        }
        return sum.divide(BigDecimal.valueOf(period), 2, BigDecimal.ROUND_HALF_UP);
    }

    private static BigDecimal calculateSupport(List<Candle> candles, int index, int window) {
        BigDecimal min = candles.get(index).low;
        for (int i = index - window + 1; i <= index; i++) {
            min = min.min(candles.get(i).low);
        }
        return min;
    }

    private static BigDecimal calculateResistance(List<Candle> candles, int index, int window) {
        BigDecimal max = candles.get(index).high;
        for (int i = index - window + 1; i <= index; i++) {
            max = max.max(candles.get(i).high);
        }
        return max;
    }

    private static void placeOrder(String symbol, String side, double amount, int leverage) {
        // Logic to place an order through the exchange API (not shown)
        System.out.println("Placing order: " + side + " " + amount + " of " + symbol + " with leverage " + leverage);
    }
}

class Candle {
    String timestamp;
    BigDecimal open;
    BigDecimal high;
    BigDecimal low;
    BigDecimal close;
    BigDecimal volume;
    BigDecimal sma125;
    BigDecimal sma125Slope;
    BigDecimal support;
    BigDecimal resistance;
    int signal;

    // Constructor and other methods (e.g., toString) can be added as needed
}
